File: bench/suite/batch/c/batch.c
/* heavy/batch in C: mmap, one pthread per core over newline-aligned byte
 * ranges, open-addressing hash tables per thread, merged at the end.
 * See bench/SPEC.md.
 *
 *   cc -O3 -march=native -flto -o batch batch.c -lpthread
 *   ./batch <file.csv>
 */
#define _GNU_SOURCE
#include <fcntl.h>
#include <inttypes.h>
#include <pthread.h>
#include <sched.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <unistd.h>

/* ---- u64 -> i64 table (user_id -> revenue) ------------------------- */

typedef struct {
    uint64_t *keys; /* key + 1, so 0 marks an empty slot */
    int64_t *vals;
    size_t cap, len;
} utab_t;

static inline uint64_t mix64(uint64_t x) {
    x ^= x >> 33;
    x *= 0xff51afd7ed558ccdULL;
    x ^= x >> 33;
    x *= 0xc4ceb9fe1a85ec53ULL;
    return x ^ (x >> 33);
}

static void utab_init(utab_t *t, size_t cap) {
    t->cap = cap;
    t->len = 0;
    t->keys = calloc(cap, sizeof *t->keys);
    t->vals = calloc(cap, sizeof *t->vals);
}

static void utab_add(utab_t *t, uint64_t key, int64_t v);

static void utab_grow(utab_t *t) {
    utab_t n;
    utab_init(&n, t->cap * 2);
    for (size_t i = 0; i < t->cap; i++)
        if (t->keys[i]) utab_add(&n, t->keys[i] - 1, t->vals[i]);
    free(t->keys);
    free(t->vals);
    *t = n;
}

static void utab_add(utab_t *t, uint64_t key, int64_t v) {
    if ((t->len + 1) * 2 > t->cap) utab_grow(t);
    size_t mask = t->cap - 1;
    size_t i = mix64(key) & mask;
    uint64_t k = key + 1;
    while (t->keys[i] && t->keys[i] != k) i = (i + 1) & mask;
    if (!t->keys[i]) {
        t->keys[i] = k;
        t->len++;
    }
    t->vals[i] += v;
}

/* ---- short string -> i64 table (sku -> revenue) --------------------- */

#define SKU_MAX 16

typedef struct {
    char key[SKU_MAX];
    uint8_t klen; /* 0 marks an empty slot */
    int64_t val;
} sslot_t;

typedef struct {
    sslot_t *slots;
    size_t cap, len;
} stab_t;

static inline uint64_t hash_bytes(const char *p, size_t n) {
    uint64_t h = 1469598103934665603ULL;
    for (size_t i = 0; i < n; i++) h = (h ^ (unsigned char)p[i]) * 1099511628211ULL;
    return mix64(h);
}

static void stab_init(stab_t *t, size_t cap) {
    t->cap = cap;
    t->len = 0;
    t->slots = calloc(cap, sizeof *t->slots);
}

static void stab_add(stab_t *t, const char *k, size_t n, int64_t v);

static void stab_grow(stab_t *t) {
    stab_t n;
    stab_init(&n, t->cap * 2);
    for (size_t i = 0; i < t->cap; i++)
        if (t->slots[i].klen) stab_add(&n, t->slots[i].key, t->slots[i].klen, t->slots[i].val);
    free(t->slots);
    *t = n;
}

static void stab_add(stab_t *t, const char *k, size_t n, int64_t v) {
    if ((t->len + 1) * 2 > t->cap) stab_grow(t);
    size_t mask = t->cap - 1;
    size_t i = hash_bytes(k, n) & mask;
    while (t->slots[i].klen && (t->slots[i].klen != n || memcmp(t->slots[i].key, k, n)))
        i = (i + 1) & mask;
    if (!t->slots[i].klen) {
        memcpy(t->slots[i].key, k, n);
        t->slots[i].klen = (uint8_t)n;
        t->len++;
    }
    t->slots[i].val += v;
}

/* ---- per-thread work -------------------------------------------------- */

typedef struct {
    int64_t count, qty, revenue;
    int seen;
} region_t;

typedef struct {
    const char *start, *end;
    int64_t rows;
    region_t regions[26 * 26];
    utab_t users;
    stab_t skus;
} job_t;

static inline int64_t parse_int(const char **pp) {
    const char *p = *pp;
    int64_t v = 0;
    while (*p >= '0' && *p <= '9') v = v * 10 + (*p++ - '0');
    *pp = p + 1; /* skip the separator */
    return v;
}

static void *work(void *arg) {
    job_t *j = arg;
    utab_init(&j->users, 1 << 20);
    stab_init(&j->skus, 1 << 18);
    const char *p = j->start, *end = j->end;
    while (p < end) {
        while (*p != ',') p++; /* ts */
        p++;
        int64_t user = parse_int(&p);
        const char *sku = p;
        while (*p != ',') p++;
        size_t sku_len = (size_t)(p - sku);
        p++;
        int64_t qty = parse_int(&p);
        int64_t price = parse_int(&p);
        region_t *r = &j->regions[(p[0] - 'A') * 26 + (p[1] - 'A')];
        p += 3; /* two letters and the newline */
        int64_t revenue = qty * price;
        r->count++;
        r->qty += qty;
        r->revenue += revenue;
        r->seen = 1;
        j->rows++;
        utab_add(&j->users, (uint64_t)user, revenue);
        if (sku_len > SKU_MAX) sku_len = SKU_MAX;
        stab_add(&j->skus, sku, sku_len, revenue);
    }
    return NULL;
}

/* ---- ranking ------------------------------------------------------------ */

typedef struct {
    int64_t revenue;
    uint64_t user;
} utop_t;

typedef struct {
    int64_t revenue;
    char sku[SKU_MAX + 1];
    uint8_t len;
} stop_t;

static int sku_less(const char *a, size_t an, const char *b, size_t bn) {
    int c = memcmp(a, b, an < bn ? an : bn);
    return c ? c < 0 : an < bn;
}

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "usage: %s <file.csv>\n", argv[0]);
        return 2;
    }
    int fd = open(argv[1], O_RDONLY);
    struct stat st;
    if (fd < 0 || fstat(fd, &st)) {
        perror(argv[1]);
        return 1;
    }
    size_t size = (size_t)st.st_size;
    const char *data = size ? mmap(NULL, size, PROT_READ, MAP_PRIVATE, fd, 0) : "";
    if (data == MAP_FAILED) {
        perror("mmap");
        return 1;
    }
    madvise((void *)data, size, MADV_SEQUENTIAL);

    int threads = 1;
#ifdef __linux__
    cpu_set_t set;
    CPU_ZERO(&set);
    if (sched_getaffinity(0, sizeof set, &set) == 0) threads = CPU_COUNT(&set);
#else
    threads = (int)sysconf(_SC_NPROCESSORS_ONLN);
#endif
    const char *w = getenv("WORKERS");
    if (w && atoi(w) > 0) threads = atoi(w);
    if ((size_t)threads > size / (1 << 16) + 1) threads = (int)(size / (1 << 16) + 1);

    job_t *jobs = calloc((size_t)threads, sizeof *jobs);
    pthread_t *tids = calloc((size_t)threads, sizeof *tids);
    const char *cursor = data;
    for (int i = 0; i < threads; i++) {
        const char *end = i == threads - 1 ? data + size : data + size / (size_t)threads * (size_t)(i + 1);
        if (end < cursor) end = cursor;
        while (end < data + size && end[-1] != '\n') end++;
        jobs[i].start = cursor;
        jobs[i].end = end;
        cursor = end;
        pthread_create(&tids[i], NULL, work, &jobs[i]);
    }
    for (int i = 0; i < threads; i++) pthread_join(tids[i], NULL);

    /* merge into the largest tables */
    int64_t rows = 0;
    region_t regions[26 * 26] = {{0}};
    utab_t *users = &jobs[0].users;
    stab_t *skus = &jobs[0].skus;
    for (int i = 0; i < threads; i++) {
        rows += jobs[i].rows;
        for (int r = 0; r < 26 * 26; r++) {
            regions[r].count += jobs[i].regions[r].count;
            regions[r].qty += jobs[i].regions[r].qty;
            regions[r].revenue += jobs[i].regions[r].revenue;
            regions[r].seen |= jobs[i].regions[r].seen;
        }
        if (jobs[i].users.len > users->len) users = &jobs[i].users;
        if (jobs[i].skus.len > skus->len) skus = &jobs[i].skus;
    }
    for (int i = 0; i < threads; i++) {
        utab_t *u = &jobs[i].users;
        if (u != users) {
            for (size_t k = 0; k < u->cap; k++)
                if (u->keys[k]) utab_add(users, u->keys[k] - 1, u->vals[k]);
            free(u->keys);
            free(u->vals);
        }
        stab_t *s = &jobs[i].skus;
        if (s != skus) {
            for (size_t k = 0; k < s->cap; k++)
                if (s->slots[k].klen) stab_add(skus, s->slots[k].key, s->slots[k].klen, s->slots[k].val);
            free(s->slots);
        }
    }

    utop_t ut[100];
    int nu = 0;
    for (size_t k = 0; k < users->cap; k++) {
        if (!users->keys[k]) continue;
        utop_t c = {users->vals[k], users->keys[k] - 1};
        if (nu == 100 && !(c.revenue > ut[99].revenue || (c.revenue == ut[99].revenue && c.user < ut[99].user)))
            continue;
        int at = nu < 100 ? nu++ : 99;
        while (at > 0 && (c.revenue > ut[at - 1].revenue ||
                          (c.revenue == ut[at - 1].revenue && c.user < ut[at - 1].user))) {
            ut[at] = ut[at - 1];
            at--;
        }
        ut[at] = c;
    }
    stop_t stp[10];
    int ns = 0;
    for (size_t k = 0; k < skus->cap; k++) {
        sslot_t *sl = &skus->slots[k];
        if (!sl->klen) continue;
        stop_t c = {.revenue = sl->val, .len = sl->klen};
        memcpy(c.sku, sl->key, sl->klen);
        #define AHEAD(a, b) ((a).revenue > (b).revenue || ((a).revenue == (b).revenue && sku_less((a).sku, (a).len, (b).sku, (b).len)))
        if (ns == 10 && !AHEAD(c, stp[9])) continue;
        int at = ns < 10 ? ns++ : 9;
        while (at > 0 && AHEAD(c, stp[at - 1])) {
            stp[at] = stp[at - 1];
            at--;
        }
        stp[at] = c;
    }

    char *out = NULL;
    size_t outlen = 0;
    FILE *o = open_memstream(&out, &outlen);
    fprintf(o, "rows=%" PRId64 "\n", rows);
    for (int a = 0; a < 26; a++)
        for (int b = 0; b < 26; b++) {
            region_t *r = &regions[a * 26 + b];
            if (r->seen)
                fprintf(o, "region=%c%c count=%" PRId64 " qty=%" PRId64 " revenue=%" PRId64 "\n",
                        'A' + a, 'A' + b, r->count, r->qty, r->revenue);
        }
    for (int i = 0; i < nu; i++)
        fprintf(o, "top_user rank=%d user_id=%" PRIu64 " revenue=%" PRId64 "\n", i + 1, ut[i].user, ut[i].revenue);
    for (int i = 0; i < ns; i++)
        fprintf(o, "top_sku rank=%d sku=%.*s revenue=%" PRId64 "\n", i + 1, stp[i].len, stp[i].sku, stp[i].revenue);
    fclose(o);
    fwrite(out, 1, outlen, stdout);
    return 0;
}

File: bench/suite/batch/java/Batch.java
// heavy/batch in Java: memory-mapped chunks, a platform thread per core,
// open-addressing tables with primitive keys merged at the end.
// See bench/SPEC.md.
import java.io.IOException;
import java.io.PrintStream;
import java.io.BufferedOutputStream;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.List;

public class Batch {
    static long mix(long x) {
        x ^= x >>> 33;
        x *= 0xff51afd7ed558ccdL;
        x ^= x >>> 33;
        x *= 0xc4ceb9fe1a85ec53L;
        return x ^ (x >>> 33);
    }

    /** long -> long, key+1 stored so 0 marks empty. */
    static final class LongMap {
        long[] keys, vals;
        int size;

        LongMap(int cap) { keys = new long[cap]; vals = new long[cap]; }

        void add(long key, long v) {
            if ((size + 1) * 2 > keys.length) grow();
            int mask = keys.length - 1;
            int i = (int) mix(key) & mask;
            long k = key + 1;
            while (keys[i] != 0 && keys[i] != k) i = (i + 1) & mask;
            if (keys[i] == 0) { keys[i] = k; size++; }
            vals[i] += v;
        }

        void grow() {
            long[] ok = keys, ov = vals;
            keys = new long[ok.length * 2];
            vals = new long[ok.length * 2];
            size = 0;
            for (int i = 0; i < ok.length; i++) if (ok[i] != 0) add(ok[i] - 1, ov[i]);
        }
    }

    /** sku (up to 16 bytes, packed into two longs plus a length) -> long. */
    static final class SkuMap {
        long[] a, b, vals;
        byte[] lens;
        int size;

        SkuMap(int cap) { a = new long[cap]; b = new long[cap]; vals = new long[cap]; lens = new byte[cap]; }

        void add(long ka, long kb, int len, long v) {
            if ((size + 1) * 2 > a.length) grow();
            int mask = a.length - 1;
            int i = (int) mix(ka * 31 + kb + len) & mask;
            while (lens[i] != 0 && (lens[i] != len || a[i] != ka || b[i] != kb)) i = (i + 1) & mask;
            if (lens[i] == 0) { a[i] = ka; b[i] = kb; lens[i] = (byte) len; size++; }
            vals[i] += v;
        }

        void grow() {
            long[] oa = a, ob = b, ov = vals;
            byte[] ol = lens;
            a = new long[oa.length * 2]; b = new long[oa.length * 2]; vals = new long[oa.length * 2]; lens = new byte[oa.length * 2];
            size = 0;
            for (int i = 0; i < oa.length; i++) if (ol[i] != 0) add(oa[i], ob[i], ol[i], ov[i]);
        }
    }

    static final class Part {
        long rows;
        final long[] regions = new long[676 * 3];
        LongMap users = new LongMap(1 << 20);
        SkuMap skus = new SkuMap(1 << 17);
    }

    static void work(FileChannel ch, long start, long end, Part p) throws IOException {
        MappedByteBuffer buf = ch.map(FileChannel.MapMode.READ_ONLY, start, end - start);
        int n = (int) (end - start), i = 0;
        while (i < n) {
            while (buf.get(i) != ',') i++;
            i++;
            long user = 0;
            byte c;
            while ((c = buf.get(i)) != ',') { user = user * 10 + (c - '0'); i++; }
            i++;
            long ka = 0, kb = 0;
            int len = 0;
            while ((c = buf.get(i)) != ',') {
                if (len < 8) ka |= (c & 0xffL) << (8 * len);
                else if (len < 16) kb |= (c & 0xffL) << (8 * (len - 8));
                len++;
                i++;
            }
            i++;
            long qty = 0;
            while ((c = buf.get(i)) != ',') { qty = qty * 10 + (c - '0'); i++; }
            i++;
            long price = 0;
            while ((c = buf.get(i)) != ',') { price = price * 10 + (c - '0'); i++; }
            i++;
            int r = ((buf.get(i) - 'A') * 26 + (buf.get(i + 1) - 'A')) * 3;
            i += 3;
            long rev = qty * price;
            p.regions[r]++;
            p.regions[r + 1] += qty;
            p.regions[r + 2] += rev;
            p.rows++;
            p.users.add(user, rev);
            p.skus.add(ka, kb, Math.min(len, 16), rev);
        }
    }

    static String sku(long a, long b, int len) {
        StringBuilder s = new StringBuilder(len);
        for (int i = 0; i < len; i++) s.append((char) (i < 8 ? (a >>> (8 * i)) & 0xff : (b >>> (8 * (i - 8))) & 0xff));
        return s.toString();
    }

    public static void main(String[] args) throws Exception {
        Path path = Path.of(args[0]);
        try (FileChannel ch = FileChannel.open(path, StandardOpenOption.READ)) {
            long size = ch.size();
            int workers = Integer.parseInt(System.getenv().getOrDefault("WORKERS",
                    String.valueOf(Runtime.getRuntime().availableProcessors())));
            // a mapping is limited to 2GB: at least one chunk per GB
            int chunks = (int) Math.max(workers, size / (1L << 30) + 1);
            chunks = (int) Math.min(chunks, size / 65536 + 1);
            long[] bounds = new long[chunks + 1];
            java.nio.ByteBuffer probe = java.nio.ByteBuffer.allocate(4096);
            for (int w = 1; w < chunks; w++) {
                long at = size / chunks * w - 1, found = size;
                outer:
                while (at < size) {
                    probe.clear();
                    int got = ch.read(probe, at);
                    if (got <= 0) break;
                    for (int k = 0; k < got; k++) {
                        if (probe.get(k) == '\n') { found = at + k + 1; break outer; }
                    }
                    at += got;
                }
                bounds[w] = Math.max(found, bounds[w - 1]);
            }
            bounds[chunks] = size;

            Part[] parts = new Part[chunks];
            List<Thread> threads = new ArrayList<>();
            java.util.concurrent.atomic.AtomicInteger next = new java.util.concurrent.atomic.AtomicInteger();
            for (int t = 0; t < Math.min(workers, chunks); t++) {
                Thread th = new Thread(() -> {
                    int c;
                    while ((c = next.getAndIncrement()) < parts.length) {
                        Part p = new Part();
                        try { work(ch, bounds[c], bounds[c + 1], p); } catch (IOException e) { throw new RuntimeException(e); }
                        parts[c] = p;
                    }
                });
                th.start();
                threads.add(th);
            }
            for (Thread th : threads) th.join();

            Part total = parts[0];
            for (int c = 1; c < parts.length; c++) {
                Part p = parts[c];
                total.rows += p.rows;
                for (int k = 0; k < total.regions.length; k++) total.regions[k] += p.regions[k];
                for (int k = 0; k < p.users.keys.length; k++) if (p.users.keys[k] != 0) total.users.add(p.users.keys[k] - 1, p.users.vals[k]);
                for (int k = 0; k < p.skus.a.length; k++) if (p.skus.lens[k] != 0) total.skus.add(p.skus.a[k], p.skus.b[k], p.skus.lens[k], p.skus.vals[k]);
                parts[c] = null;
            }

            long[] topRev = new long[100], topUser = new long[100];
            int nu = 0;
            LongMap u = total.users;
            for (int k = 0; k < u.keys.length; k++) {
                if (u.keys[k] == 0) continue;
                long rev = u.vals[k], user = u.keys[k] - 1;
                if (nu == 100 && !(rev > topRev[99] || (rev == topRev[99] && user < topUser[99]))) continue;
                int at = nu < 100 ? nu++ : 99;
                while (at > 0 && (rev > topRev[at - 1] || (rev == topRev[at - 1] && user < topUser[at - 1]))) {
                    topRev[at] = topRev[at - 1];
                    topUser[at] = topUser[at - 1];
                    at--;
                }
                topRev[at] = rev;
                topUser[at] = user;
            }
            long[] sRev = new long[10];
            String[] sName = new String[10];
            int ns = 0;
            SkuMap s = total.skus;
            for (int k = 0; k < s.a.length; k++) {
                if (s.lens[k] == 0) continue;
                long rev = s.vals[k];
                if (ns == 10 && rev < sRev[9]) continue;
                String name = sku(s.a[k], s.b[k], s.lens[k]);
                if (ns == 10 && !(rev > sRev[9] || name.compareTo(sName[9]) < 0)) continue;
                int at = ns < 10 ? ns++ : 9;
                while (at > 0 && (rev > sRev[at - 1] || (rev == sRev[at - 1] && name.compareTo(sName[at - 1]) < 0))) {
                    sRev[at] = sRev[at - 1];
                    sName[at] = sName[at - 1];
                    at--;
                }
                sRev[at] = rev;
                sName[at] = name;
            }

            PrintStream out = new PrintStream(new BufferedOutputStream(System.out, 1 << 16), false);
            out.print("rows=" + total.rows + "\n");
            for (int c = 0; c < 676; c++) {
                if (total.regions[c * 3] == 0) continue;
                out.print("region=" + (char) ('A' + c / 26) + (char) ('A' + c % 26) + " count=" + total.regions[c * 3]
                        + " qty=" + total.regions[c * 3 + 1] + " revenue=" + total.regions[c * 3 + 2] + "\n");
            }
            for (int i = 0; i < nu; i++) out.print("top_user rank=" + (i + 1) + " user_id=" + topUser[i] + " revenue=" + topRev[i] + "\n");
            for (int i = 0; i < ns; i++) out.print("top_sku rank=" + (i + 1) + " sku=" + sName[i] + " revenue=" + sRev[i] + "\n");
            out.flush();
        }
    }
}
